app/assets/javascripts/components/ppm-estimator.js

```diff
@@ -0,0 +1,162 @@
+
+      // The visibility of the spouse progear field is controlled elsewhere, so it is safe to make this label visible regardless
+      var proGearWeightSpouse = getEntitlementProgearSpouse(entitlement);
+      $('#entitlement_progear_spouse').text(proGearWeightSpouse.toString());
+      $progearSpouseAllowanceText.css('visibility', 'visible');
+      onWeightProgearSpouseInput(null);
+    }
+  };
+
+  var validateEntitlementField = function($input, getEntitlementFn) {
+    // an empty field is never an error
+    if ($input.val() == '')
+      return true;
+
+    // Some kinds of input validation can happen whether or not we know what the entitlement is
+    var weightLbs = parseInt($input.val());
+    if (weightLbs == NaN || weightLbs < 0)
+      return false;
+
+    // If the rank has not been chosen yet, then the entitlement cannot be loaded. Whatever the number is, it's fine for now
+    if (!$rank.val())
+      return true;
+
+    var entitlement = entitlementsJson.find(function(data) { return data.slug == $rank.val(); });
+    var entitlementWeight = getEntitlementFn(entitlement);
+    return weightLbs <= entitlementWeight;
+  };
+
+  var onWeightInput = function(e) {
+    if (validateEntitlementField($('#weight'), getEntitlementWeightSelf))
+      $('#weight-section').removeClass('usa-input-error');
+    else
+      $('#weight-section').addClass('usa-input-error');
+  };
+
+  var onWeightProgearInput = function(e) {
+    if (validateEntitlementField($('#weight_progear'), getEntitlementProgear))
+      $('#progear-section').removeClass('usa-input-error');
+    else
+      $('#progear-section').addClass('usa-input-error');
+  };
+
+  var onWeightProgearSpouseInput = function(e) {
+    if (validateEntitlementField($('#weight_progear_spouse'), getEntitlementProgearSpouse))
+      $('#progear-spouse-section').removeClass('usa-input-error');
+    else
+      $('#progear-spouse-section').addClass('usa-input-error');
+  };
+
+  $rank.on('change', onPersonalInfoChanged);
+  $('#dependents_yes').change(onPersonalInfoChanged);
+  $('#dependents_no').change(onPersonalInfoChanged);
+  $('#married_yes').change(function() { $('#progear-spouse-section').removeAttr('hidden'); });
+  $('#married_no').change(function() { $('#progear-spouse-section').attr('hidden', true); });
+
+  $('#weight').on('input', onWeightInput);
+  $('#weight_progear').on('input', onWeightProgearInput);
+  $('#weight_progear_spouse').on('input', onWeightProgearSpouseInput);
+
+  var onZipCodeKey = function(evt) {
+    var charCode = (evt.which) ? evt.which : event.keyCode
+    // charCode 8 is backspace, required for Firefox to allow backspace when there are already 5 numbers in the box
+    return evt.target.value.length < 5 || charCode == 8;
+  };
+
+  $('#start').keypress(onZipCodeKey);
+  $('#end').keypress(onZipCodeKey);
+
+})(window, jQuery);
+
+function onEditDetails() {
+  $('#ppm-estimate-form').removeAttr('hidden');
+  $('#ppm-estimate-results').attr('hidden', true);
+}
```

Comment:
I'd like to see the miscellaneous functions and functionality in this file wrapped up into a "class," either `PpmEstimator` or one or more groupings that better map to the various interactive components on the PPM Estimator page.

I'm not familiar enough with the UI to know what that might look like in practice, though. Let me know if you want some help with that.